compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -11,19 +11,30 @@ use rustc_span::symbol::{kw, sym};
 use rustc_span::Span;
 use tracing::debug;
 
+enum ImplTraitKind {
+    Existential,
+    Universal(LocalDefId),
+}
+
 crate fn collect_definitions(
     resolver: &mut Resolver<'_>,
     fragment: &AstFragment,
     expansion: ExpnId,
 ) {
     let parent_def = resolver.invocation_parents[&expansion];
-    fragment.visit_with(&mut DefCollector { resolver, parent_def, expansion });
+    fragment.visit_with(&mut DefCollector {
+        resolver,
+        parent_def,
+        impl_trait: ImplTraitKind::Existential,
```

Comment:
I'm pretty sure this is incorrect, `collect_definitions` is called on an arbitrary AST fragment (expanded from a macro), not on crate root, and that fragment can be in any "impl-trait context".

You'll have to track "impl-trait contexts" associated with individual macro expansions and pass them to `collect_definitions` from the outside to do this correctly.

Reply:
We are already doing a very similar bookkeeping for expansions' parent def-ids, see `Resolver::invocation_parents`.
Parent impl-trait contexts can be kept in the same table.

Feel free to factor out the "Make def_key and HIR parenting consistent" change to a separate PR and assign it to me.